services/cd-service/pkg/mapper/environments_config_test.go

```diff
@@ -0,0 +1,380 @@
+						makeEnv(nameDevDe, nameDevDe, makeUpstreamLatest(), 0, api.Priority_PROD),
+					},
+					DistanceToUpstream: 0,
+				},
+			},
+		},
+		{
+			Name: "Two Environments are two Groups",
+			InputEnvs: map[string]config.EnvironmentConfig{
+				nameDevDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Latest: true,
+					},
+					ArgoCd: nil,
+				},
+				nameStagingDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Environment: nameDevDe,
+					},
+					ArgoCd: nil,
+				},
+			},
+			ExpectedResult: []*api.EnvironmentGroup{
+				{
+					EnvironmentGroupName: nameDevDe,
+					Environments: []*api.Environment{
+						makeEnv(nameDevDe, nameDevDe, makeUpstreamLatest(), 0, api.Priority_PRE_PROD),
+					},
+					DistanceToUpstream: 0,
+				},
+				{
+					EnvironmentGroupName: nameStagingDe,
+					Environments: []*api.Environment{
+						makeEnv(nameStagingDe, nameStagingDe, makeUpstreamEnvironment(nameDevDe), 1, api.Priority_PROD),
+					},
+					DistanceToUpstream: 1,
+				},
+			},
+		},
+		{
+			// note that this is not a realistic example, we just want to make sure it does not crash!
+			// some outputs may be nonsensical (like distanceToUpstream), but that's fine as long as it's stable!
+			Name: "Two Environments with a loop",
+			InputEnvs: map[string]config.EnvironmentConfig{
+				nameDevDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Environment: nameStagingDe,
+					},
+					ArgoCd: nil,
+				},
+				nameStagingDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Environment: nameDevDe,
+					},
+					ArgoCd: nil,
+				},
+			},
+			ExpectedResult: []*api.EnvironmentGroup{
+				{
+					EnvironmentGroupName: nameDevDe,
+					Environments: []*api.Environment{
+						makeEnv(nameDevDe, nameDevDe, makeUpstreamEnvironment(nameStagingDe), 667, api.Priority_PRE_PROD),
+					},
+					DistanceToUpstream: 667,
+				},
+				{
+					EnvironmentGroupName: nameStagingDe,
+					Environments: []*api.Environment{
+						makeEnv(nameStagingDe, nameStagingDe, makeUpstreamEnvironment(nameDevDe), 668, api.Priority_PROD),
+					},
+					DistanceToUpstream: 668,
+				},
+			},
+		},
+		{
+			// note that this is not a realistic example, we just want to make sure it does not crash!
+			// some outputs may be nonsensical (like distanceToUpstream), but that's fine as long as it's stable!
+			Name: "Two Environments with non exists upstream",
+			InputEnvs: map[string]config.EnvironmentConfig{
+				nameDevDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Latest: true,
+					},
+					ArgoCd: nil,
+				},
+				nameStagingDe: {
+					Upstream: &config.EnvironmentConfigUpstream{
+						Environment: nameWhoKnows,
+					},
+					ArgoCd: nil,
+				},
+			},
+			ExpectedResult: []*api.EnvironmentGroup{
+				{
+					EnvironmentGroupName: nameDevDe,
+					Environments: []*api.Environment{
+						makeEnv(nameDevDe, nameDevDe, makeUpstreamLatest(), 0, api.Priority_UPSTREAM),
```

Comment:
This will fail because the other environmentGroupName has a non existent upstream, correct?

Reply:
No this will not fail. Technically, if the real environment setup in the manifest repo was like that, then yes! it will log a warning/fail. 

But this function does not check or validate that, it just assigns a random distance i.e. 667  We don't want this to crash / panic here.